[Lee Yi Heng] iP

leeyiheng12 opened the pull request:
# Super cool version of Duke!

## Why use it?
- It's very likely error-free!! :upside_down_face:
> Just Do It - Nike

## How to use it:
1. Install Java
2. Click [here](https://github.com/leeyiheng12/ip)
3. Check the Releases section at the side to find a .jar file

## How to get the source code
```cmd
C:\Users\damith> git clone https://github.com/leeyiheng12/ip.git
```

## To-do
- [x] Have Java installed
- [ ] Have Git installed
- [ ] `git clone` my repo

Commit:
Add Gradle support

leeyiheng12 committed:
lvl1: made the bot echo user input

leeyiheng12 committed:
added list functionality, changed name of helper method for printing

leeyiheng12 committed:
add functionality to mark task as done or undone

leeyiheng12 committed:
added support for different types of tasks

leeyiheng12 committed:
added testing functionality

leeyiheng12 committed:
split classes into different files, added error handling

leeyiheng12 committed:
added support for deleting tasks

leeyiheng12 committed:
added enum for type of task

leeyiheng12 committed:
included exception throwing to handle errors

leeyiheng12 committed:
add save functionality to a file stored locally

leeyiheng12 committed:
allow duke to understand and parse datetime formats

leeyiheng12 committed:
Merge branch 'branch-Level-8'

leeyiheng12 committed:
refactor code into classes such as Storage, Ui, Parser, Tasklist

leeyiheng12 committed:
divide files into packages, other small changes

leeyiheng12 committed:
add unit tests for parser and tasklist

leeyiheng12 committed:
package Duke as .jar, minor changes

leeyiheng12 committed:
add javadocs comments to classes and methods

leeyiheng12 committed:
reformat code to match coding standard

leeyiheng12 committed:
add functionality to let a user search for a keyword

leeyiheng12 committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/Duke.java
#	src/main/java/storage/Storage.java
#	src/main/java/tasks/Task.java
#	src/main/java/user/Ui.java

leeyiheng12 committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/tasks/Deadline.java
#	src/main/java/tasks/Event.java
#	src/main/java/tasks/Task.java
#	src/main/java/tasks/Todo.java
#	src/main/java/user/Parser.java
#	src/main/java/user/Ui.java

leeyiheng12 committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle

leeyiheng12 committed:
integrate checkstyle to detect coding style violations, and fix violations

leeyiheng12 committed:
add a gui to duke

leeyiheng12 committed:
use varargs for prettyprint function

hqhqhq1 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,51 @@
+        String userInput;
+        String[] result;
+        boolean finished = false;
+        while (!finished) {
```
maybe can put all this in a method call run() and the main method just evoke run() method

hqhqhq1 commented on src/main/java/tasks/Task.java:
```diff
@@ -0,0 +1,74 @@
+    public static final String INVALID_TASK_NUM_ERROR_STRING = "Not a valid task number!";
+    public static final String TASK_NUM_DOES_NOT_EXIST_ERROR_STRING = "Task %d does not exist!";
+
+    public enum TaskType {
```
Why not add "delete, find" as a TaskType too

hqhqhq1 commented on src/main/java/user/Ui.java:
```diff
@@ -0,0 +1,184 @@
+            + INDENT + "|____/ \\__,_|_|\\_\\___|\n";
+    private static final String[] OPENING_MESSAGE = new String[] {"Hello! I'm Duke! What can I do for you?"};
+    private static final String CLOSING_MESSAGE = "Bye. Hope to see you again soon!";
+    private static final String[] helpStrings = new String[] {
```
I like this user guide 👍

leeyiheng12 committed:
update gradle build file

leeyiheng12 committed:
add a couple of assertions, delete old code

leeyiheng12 committed:
improve code quality

leeyiheng12 committed:
Merge pull request #2 from leeyiheng12/branch-A-CodeQuality

improve code quality

leeyiheng12 committed:
Merge branch 'master' into branch-A-Assertions

leeyiheng12 committed:
Merge pull request #1 from leeyiheng12/branch-A-Assertions

add a couple of assertions, delete old code

tonishka commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,52 @@
+        Duke duke = new Duke();
+        duke.ui.greet();
+        String userInput;
+        boolean finished = false;
```
```suggestion
        boolean isFinished = false;
```

tonishka commented on src/main/java/tasks/Deadline.java:
```diff
@@ -0,0 +1,66 @@
+    private static final DateTimeFormatter INPUT_FORMAT = DateTimeFormatter.ofPattern("dd/MM/yyyy-HH:mm");
+    private static final DateTimeFormatter PRETTY_FORMAT = DateTimeFormatter.ofPattern("MMM dd yyyy HH:mm");
+
+    private final Task.TaskType taskType = Task.TaskType.DEADLINE;
+    private final String taskName;
+    private final LocalDateTime deadline;
```
As they are constants, shouldn't they be in upper case with underscores separating different words?

tonishka commented on src/main/java/tasks/Event.java:
```diff
@@ -0,0 +1,66 @@
+    private static final DateTimeFormatter INPUT_FORMAT = DateTimeFormatter.ofPattern("dd/MM/yyyy-HH:mm");
+    private static final DateTimeFormatter PRETTY_FORMAT = DateTimeFormatter.ofPattern("MMM dd yyyy HH:mm");
+
+    private final Task.TaskType taskType = Task.TaskType.EVENT;
+    private final String taskName;
+    private final LocalDateTime eventTime;
```
Since these variables are constants, shouldn’t they be in uppercase?
```suggestion
    private final Task.TaskType TASK_TYPE = Task.TaskType.EVENT;
    private final String TASK_NAME;
    private final LocalDateTime EVENT_TIME;
```

tonishka commented on src/main/java/tasks/Task.java:
```diff
@@ -0,0 +1,73 @@
+    }
+    private boolean isDone = false;
+
+    /**
+     * Sets the Task as done or undone.
+     *
+     * @param newDone True if the task is to be set as done, false to be set as undone.
+     */
```
I appreciate the extensive use of JavaDoc comments in your code, but I think since this a trivial setter method it does not warrant a JavaDoc comment of its own?

tonishka commented on src/main/java/user/Ui.java:
```diff
@@ -0,0 +1,174 @@
+            + INDENT + "|____/ \\__,_|_|\\_\\___|\n";
+    private static final String[] OPENING_MESSAGE = new String[] {"Hello! I'm Duke! What can I do for you?"};
+    private static final String CLOSING_MESSAGE = "Bye. Hope to see you again soon!";
+    private static final String[] helpStrings = new String[] {
+        "Commands:",
+        "'todo [some activity]' - add a todo",
+        "'deadline [some task] /by [dd/mm/yyyy-hh:mm]' - add a deadline",
+        "'event [some event] /at [dd/mm/yyyy-hh:mm]' - add an event",
+        "'list'- view all tasks",
+        "'find [keyword]' - search tasks by keyword",
+        "'mark [task number]' - mark a task as done",
+        "'unmark [task number]' - mark a task as not done",
+        "'delete [task number]' - delete a task"
```
User guide is very comprehensive and easy to understand

tonishka commented on src/main/java/tasks/Task.java:
```diff
@@ -0,0 +1,73 @@
+/** A class that functions as an abstraction of a task. */
+public abstract class Task {
+
+    public static final String UNKNOWN_INPUT_ERROR_STRING = "I don't think I know what this is!";
+    public static final String BAD_DESCRIPTION_ERROR_STRING = "Description cannot be empty!";
+    public static final String INVALID_TASK_NUM_ERROR_STRING = "Not a valid task number!";
+    public static final String TASK_NUM_DOES_NOT_EXIST_ERROR_STRING = "Task %d does not exist!";
```
I like how you have overall handled errors and exceptions in your classes 👍

tonishka reviewed (commented):
All in all, I found your code very readable and well organised. I really like how you have handled exceptions in Duke - it covers a wide range of user errors and is abstracted away nicely in each class. Found a couple of minor coding standard violations, but looks good overall 👍

leeyiheng12 committed:
add functionality to support loan records

leeyiheng12 committed:
Make small quality-of-life changes here and there.

The most important change is to handle the case where the user
enters a borrowing or lending amount, but entered characters other
than 0-9.

A title is also added to the Duke GUI.

leeyiheng12 committed:
Improve the GUI.

Duke is changed to HAL 4500, akin to HAL 9000 from the film
2001: A Space Odyssey. Background colors are added to text boxes.
Errors are shown in red.

leeyiheng12 committed:
add user guide and screenshot

leeyiheng12 committed:
Make final changes to the bot.

A button is available to show the link to the GitHub repo.
Some Java files are now in a package for better organization.
There are small changes to pieces of code here and there.

leeyiheng12 committed:
Edit the screenshot of the UI to reflect the latest change.

leeyiheng12 committed:
delete useless code from when it was a CLI application

leeyiheng12 committed:
Decrease height of window to accommodate smaller screen sizes
